include/spinlock.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef _SPINLOCK_H
+#define _SPINLOCK_H
+
+#include <atomic.h>
+
+struct k_spinlock_key {
+	int key;
+};
+
+typedef struct k_spinlock_key k_spinlock_key_t;
+
+struct k_spinlock {
+#ifdef CONFIG_SMP
+	atomic_t locked;
+#ifdef CONFIG_DEBUG
+	int saved_key;
+#endif
+#endif
+};
+
+static inline k_spinlock_key_t k_spin_lock(struct k_spinlock *l)
+{
+	k_spinlock_key_t k;
+
+	/* Note that we need to use the underlying arch-specific lock
+	 * implementation.  The "irq_lock()" API in SMP context is
+	 * actually a wrapper for a global spinlock!
+	 */
+	k.key = _arch_irq_lock();
+
+#ifdef CONFIG_SMP
+# ifdef CONFIG_DEBUG
+	l->saved_key = k.key;
+# endif
+	while (!atomic_cas(&l->locked, 0, 1)) {
```

Comment:
I'm not sure if it would be an issue in the systems Zephyr runs on (esp. on ESP32, where this makes sense now), but it's often a good thing to perform the locking primitive of a spinlock like this in order to reduce cache contention:

```C
while (!atomic_cas(&l->locked, 0, 1)) {
     while (l->locked != 0) {
     }
}
```

Note that `atomic_get()` is not used here on purpose: the generic implementation using GCC atomic builtins will enforce total ordering, which isn't needed because `atomic_cas()` will do that for us if the value ever changes.

Reply:
Even on x86, the resulting lock cmpxchg is safe from cache spamming.  The line goes into a S(hared) state in L1D the first time through the loop and the core hits L2/L3 only if it snoops someone else's change to that cache line.

But really long term if we care about performance what we'll do is to have each architecture code its own spinlock implementation with inlines instead of calling out to the generic atomic API.